### DISCUSSION TITLE
过期：RIFE 4k scale=0.5

### DISCUSSION THREAD

**Discussion by ww2283** [hooke007/mpv_PlayKit#235] [回收]
请问如果我想试试设置RIFE scale=0.5来做4k插帧的话，修改下面这个量对吗？

cut3 = RIFE(clip=cut2, multi=FPS_num, model=RIFEModel.v4_6, backend=Backend.TRT(fp16=True, scale="0.5", device_id=GPU, use_cuda_graph=True, num_streams=GPU_t))

这是rife_cuda.vpy里面的。目前的情况是如果这么改，没有任何插帧，输出FPS就是原本的FPS。谢谢！

**Comment by hooke007** [comment]
报错已经显示在控制台了，去看一眼就知道了。
只改脚本没用，缺numpy和onnx依赖
只改scale没适用性，切片大小也变了

**Reply by ww2283** [reply]
哦好吧，谢谢！用RIFE单独做4k插帧的时候scale=0.5还是可以进行的，就想试试实时可不可以。那就这样吧～

**Reply by hooke007** [reply]
你说的应该不是一个东西，这个“scale”是在vsmlrt的wrapper里定义的。

**Reply by ww2283** [reply]
这个目前我比较晕。。rife自己的script的话是在这里。
https://github.com/megvii-research/ECCV2022-RIFE/blob/58f23dc85a7f573157fa0a5447cee38742353c7e/inference_video.py#L64

**Reply by hooke007** [reply]
那个是rife自己的参考实现；
把rife用到mpv里靠的是vapoursynth滤镜支持的插件库

**Reply by ww2283** [reply]
哦好的，这个我大概明白了～

**Reply by hooke007** [reply]
https://github.com/hooke007/MPV_lazy/commit/715ab00fbecbb3eee43ad7cc5b43eb20f03c0999 当然依然需要onnx

**Reply by ww2283** [reply]
这个赞！我实验实验，目前playback有点慢，像是半速播放，不过确实能播放

**Reply by hooke007** [reply]
https://github.com/hooke007/MPV_lazy/commit/91d4c2ce1f5f1f672dd7eab30f2aec8ad96225e7 实测下来还是没啥明显改进

> 像是半速播放

看一眼stats的丢帧数就知道了，渲染队列跟不上

**Reply by ww2283** [reply]
不过我现在还是卡在onnx报错上：modulenotfounderror no module named onnx
尬了。。我按照https://onnxruntime.ai/docs/install/#python-installs用administrator cmd.exe装好了onnx，可是无论如何控制台都提示onnx没有安装。。

**Reply by hooke007** [reply]
不使用 onnxruntime 用的是 onnx

**Reply by ww2283** [reply]
(https://onnxruntime.ai/docs/install/#install-onnx-to-export-the-model) 这里说ONNX is built into PyTorch。
然后我也能看到Requirement already satisfied: onnx>=1.2.1 in c:\program files\python37\lib\site-packages (from skl2onnx) (1.12.0)。
是不是我还是miss了什么。。？

**Reply by hooke007** [reply]
如果你是自己搭建的vs环境不是用的我的话，那错误太多了。我的是基于py3.10，而且vs当前的版本不支持py3.7

**Reply by ww2283** [reply]
C:\Windows\System32>pip install onnx
Requirement already satisfied: onnx in c:\program files\python37\lib\site-packages (1.12.0)
Requirement already satisfied: typing-extensions>=3.6.2.1 in c:\program files\python37\lib\site-packages (from onnx) (4.4.0)
Requirement already satisfied: numpy>=1.16.6 in c:\program files\python37\lib\site-packages (from onnx) (1.21.6)
Requirement already satisfied: protobuf<=3.20.1,>=3.12.2 in c:\program files\python37\lib\site-packages (from onnx) (3.20.1)

挠头。。
哦对了，报告一个稍微有点用的信息：2080ti的水平是正好1080p 60FPS+KrigBilateral+FSRCNNX_x2_8_0_4_1+SSimDownscaler吃满，72FPS的话会掉帧，120FPS 0.5倍速的话不掉帧。所以我估计，如果是4090的话，不开scale=0.5的话（开了也不一定能成）应该做不到4k 60FPS。

**Reply by hooke007** [reply]
vs最后一个支持py3.7的版本是R48，你确定你在用这个远古的版本吗

**Reply by ww2283** [reply]
哦我没打算自己整一个vs环境，我用的完全是懒人包的，那说明我还是搞错了。。请问怎么能让懒人包的vs识别到onnx呢？我需要build一个onnx（基于https://github.com/onnx/onnx#build-onnx-from-source）然后放在懒人包的某个地方吗？

**Reply by hooke007** [reply]
直接在懒人包里的python环境安装onnx，前提是先装pip

**Comment by ww2283** [comment]
<img width="280" alt="Screenshot 2022-11-27 142459" src="https://user-images.githubusercontent.com/21956104/204155546-fdec478b-e117-4692-9593-10f2577ced0a.png">
好的我似乎是成功安装了，在mpv-lazy\lib\site-packages\能看到这些已经安装的package了。我还需要在哪个文件中写上这个path吗？目前控制台还是提示没找到onnx

**Reply by hooke007** [reply]
https://gist.github.com/hooke007/e502688c60ef1c5f9ef507cf4db25b34#21-%E5%AE%89%E8%A3%85-pip-%E5%8F%AF%E9%80%89

**Reply by ww2283** [reply]
感谢！成功了！果然2080TI对于4k还是勉强了，2x插帧也还是掉帧，不过scale=0.5似乎解决了显存不足的问题，目前11G显存也够用了~

**Reply by hooke007** [reply]
有没有用新版测试过4k下的表现，最好是sdr8bit的片源（减少干预变量）
优先使用静态引擎

也可以把这个脚本放到mpv.exe的位置，然后此处打开终端执行
<s>[rife_t.zip](https://github.com/hooke007/MPV_lazy/files/10408961/rife_t.zip)</s>
```ps
./vspipe -p rife_t.vpy .
```
可以测出理论的4k速度，排除了mpv中的其它瓶颈干扰

**Reply by ww2283** [reply]
卧槽，这个不错，这就科学多了，回头我试试来报告，这两天有点忙
简单来说（目前主观），4k 8bit有进步，虽然也是疯狂跳帧，但是勉强能看（非常勉强）。也算是很大进步了，要知道2080ti上一个版本的rife_cuda是卡到几乎动不了的。4k 10bit没有明显改善。1080p在上一个版本就没有什么问题，这次GPU占用下来了10%，而且如果用上下箭头快速跳转的话，很流畅，几乎没有卡顿。

**Reply by hooke007** [reply]
随意何时测试。我重写了上面的脚本想验证一些事情，那个就不用看了，后边会单独开一贴子
- https://github.com/hooke007/MPV_lazy/discussions/255#discussioncomment-4685398
试试 全开 Cuda_graph, Cublas, Cudnn 三项和默认下的分数差异。

**Reply by ww2283** [reply]
windows系统我是实在太不灵了。。
我先到mpv.exe目录下，右键打开terminal，是个powershell，然后，这个怎么破？
PS J:\mpv-lazy> ./vspipe -p BM_rife.vpy .
CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage. See `CUDA_MODULE_LOADING` in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage. See `CUDA_MODULE_LOADING` in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
Failed to retrieve output node. Invalid index specified?

**Reply by hooke007** [reply]
额你更新文件了吗（

**Reply by ww2283** [reply]
更了。。吧？是1.11号重新下载的，先下载了release，然后打上rifeCUDA的包，最后用main里面的文件替换了release里面的。目前看片子是没问题的。还有别的更新吗？

**Reply by ww2283** [reply]
到这个地方都是成功的：
[01/18/2023-14:40:33] [I] Saved 244899 bytes of timing cache to J:/mpv-lazy/vapoursynth64/plugins\models\rife\rife_v4.6.onnx.1920x1088_fp16_workspace256_trt-8501_I-fp16_O-fp16_NVIDIA-GeForce-RTX-2080-Ti_3dcbe72f.engine.cache
[01/18/2023-14:40:33] [I] Engine built in 166.668 sec.

**Reply by hooke007** [reply]
额。。我说的是上游mlrt的包

**Reply by ww2283** [reply]
按照https://gist.github.com/hooke007/e502688c60ef1c5f9ef507cf4db25b34，最后一步有错误：
PS J:\mpv-lazy> ./python
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import vapoursynth as vs
>>> from vapoursynth import core
>>> print(core. version())
VapourSynth Video Processing Library
Copyright (c) 2012-2022 Fredrik Mellbin
Core R61
API R4.0
API R3.6
Options: -

>>> print(core.list_functions())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src\cython\vapoursynth.pyx", line 2434, in vapoursynth._CoreProxy.__getattr__
  File "src\cython\vapoursynth.pyx", line 2308, in vapoursynth.Core.__getattr__
AttributeError: No attribute with the name list_functions exists. Did you mistype a plugin namespace?
>>>

**Reply by hooke007** [reply]
那个有点过时了，我另外找时间看看上游改了啥。环境我另外发od上吧

**Reply by ww2283** [reply]
> 额。。我说的是上游mlrt的包

https://github.com/hooke007/MPV_lazy/discussions/123#discussioncomment-4659072 是说这个吧？1.11号的时候我装了的，不装的话vs_rife不起作用

那我等招呼了~

**Reply by hooke007** [reply]
在我release的网盘链接里，点进去有个test.7z，解压然后进目录执行刚才的命令

**Reply by hooke007** [reply]
我猜测是没通过正经pip途径安装的的onnx少了组件（

**Reply by ww2283** [reply]
这个行，Script evaluation done in 170.50 seconds
Output 4000 frames in 24.42 seconds (163.81 fps)
我想问问，从测试的数据（结合你的4080），4090能带动4k 60fps不？

**Reply by hooke007** [reply]
带动4k 60fps是指什么，60→120？

上游的wiki里有公布4090的数据(过时了

**Reply by hooke007** [reply]
>Output 4000 frames in 24.42 seconds (163.81 fps)

2080跑1080p肯定没问题啊。。2070都能跑了

我主要是想看4k下和mpv的差值

**Reply by ww2283** [reply]
哦我是说4090 能不能带动4k 30P->60P. 其实我感觉对于动画片来说，最好的方式可能是后台给做到24P/30P->120P，但屏幕输出是这120P中均匀间隔的24P/30P，不知道能不能做到。太顺滑的60P看起来有点超自然，还是有正常动作流动性的24P/30P好一些。

**Reply by hooke007** [reply]
>能不能带动4k 30P->60P

我的4080就已经可以顶格跑了。
<img src="https://user-images.githubusercontent.com/41094733/213587653-045886e0-f308-4066-926b-634666ddb7db.jpg" height=300>

24预处理到30帧的现有方案只能通过cpu处理，效率上有点问题。

> 最好的方式可能是后台给做到24P/30P->120P

mpv的 --interpolation 已经很好的完成了这部分工作了。
rife处理到48帧，interpolation对48重采样再混成到120帧匹配你的显示器的刷新率120Hz（interpolation后的帧数参考stat界面上的`Display FPS`那一行后边的estimated数据）

**Reply by ww2283** [reply]
不赖啊，多谢，过几天等4090不耍猴了我搞一个来看4k！
又学到了新知识～不过我的意思不是适配120Hz的屏或者是要看到120FPS或24->30FPS，而是把动画显得像是24FPS正常motion的电影。动画制作中有很多静态帧和静态画面的场景扫描式转换，使得播放时的有效帧率只有十几帧。rife虽然让动画更smooth了，可是毕竟不是制作组的本意。制作组只是因为成本和制作方式的原因不能做成真正的24FPS。那么理论上（我瞎想的）可以用rife做一个x5，把24FPS升到120FPS，但是在真正输出的时候只输出第1，5，10，15...120帧，这样就做到了movie motion的24FPS动画。（搓手

**Reply by hooke007** [reply]
根本原因是做不到非线性补偿吧。两点生成的中间帧永远只能取正中间的点作为中间帧（无法猜测此点此时正在做什么运动），原始信息的缺失导致无论怎样也不可能还原制作者的意图。

**Reply by ww2283** [reply]
嗯。只是非常好奇这种能是什么效果。其实稍微费点事也不是不能干：用rife先制作好120FPS的文件，然后再用ffmpeg抽帧做成24FPS的。过两天我试试看
